lib/scanner/xml-scanner/xml-scanner.h

```diff
@@ -0,0 +1,93 @@
+#include "messages.h"
+#include "logmsg/logmsg.h"
+
+#include <string.h>
+
+typedef void (*PushCurrentKeyValue)(const gchar *name, const gchar *value, gssize value_length, gpointer user_data);
+
+typedef void (*StartElement) (GMarkupParseContext *context, const gchar *element_name, const gchar **attribute_names,
+                              const gchar **attribute_values, gpointer user_data, GError **error);
+typedef void (*EndElement) (GMarkupParseContext *context, const gchar *element_name, gpointer user_data,
+                            GError **error);
+typedef void (*Text) (GMarkupParseContext *context, const gchar *text, gsize text_len, gpointer user_data,
+                      GError **error);
+
+typedef struct
+{
+  gboolean strip_whitespaces;
+  GList *exclude_tags;
+} XMLScannerOptions;
+
+typedef struct
+{
+  GMarkupParseContext *xml_ctx;
+  GMarkupParser scanner_callbacks;
+  XMLScannerOptions *options;
+  gboolean pop_next_time;
+  gboolean matchstring_shouldreverse;
+  GPtrArray *exclude_patterns;
+  PushCurrentKeyValue push_function;
+  StartElement start_element_function;
+  EndElement  end_element_function;
+  Text  text_function;
+  gpointer user_data;
+  GString *key;
+  GString *text;
+} XMLScanner;
+
+XMLScanner *xml_scanner_new(void);
+void xml_scanner_free(XMLScanner *self);
+void xml_scanner_init_instance(XMLScanner *self);
+void xml_scanner_init(XMLScanner *self, XMLScannerOptions *options, PushCurrentKeyValue push_function,
+                      gpointer user_data, gchar *key_initial);
+void xml_scanner_deinit(XMLScanner *self);
+void xml_scanner_parse(XMLScanner *self, const gchar *input, gsize input_len, GError **error);
+
+static inline void
+xml_scanner_push_current_key_value(XMLScanner *self, const gchar *name, const gchar *value, gssize value_length)
+{
+  g_assert(self->push_function); // FIXME bet way to detect programming errors?
+  self->push_function(name, value, value_length, self->user_data);
+}
+
+void xml_scanner_start_element_method(GMarkupParseContext *context, const gchar *element_name,
+                                      const gchar **attribute_names, const gchar **attribute_values, gpointer user_data, GError **error);
+void xml_scanner_end_element_method(GMarkupParseContext *context, const gchar *element_name, gpointer user_data,
+                                    GError **error);
+void xml_scanner_text_method(GMarkupParseContext *context, const gchar *text, gsize text_len, gpointer user_data,
+                             GError **error);
+
+void xml_scanner_options_set_exclude_tags(XMLScannerOptions *self, GList *exclude_tags);
+void xml_scanner_options_set_strip_whitespaces(XMLScannerOptions *self, gboolean setting);
+void xml_scanner_options_copy(XMLScannerOptions *dest, XMLScannerOptions *source);
+// BETTER NAME? no init() for options, free() is not a better name either. "clean" is stolen from csv-scanner
+void xml_scanner_options_clean(XMLScannerOptions *self);
```

Comment:
What do you think about `defaults, init, destroy and copy`?
We have this pattern in LogReader, for example.

Reply:
- clean -> destroy: ok,
- copy remains copy
- defaults added with minimal implementation
- init not going to be added as it does not have implementation